evm/src/cpu/kernel/asm/hash/blake2b/compression.asm

```diff
@@ -0,0 +1,285 @@
+global blake2b_compression:
+    // stack: retdest
+    PUSH 0
+    // stack: cur_block = 0, retdest
+    %blake2b_initial_hash_value
+compression_loop:
+    // stack: h_0, ..., h_7, cur_block, retdest
+    
+    // Store the hash values.
+    %blake2b_hash_value_addr
+    // stack: addr, h_0, ..., h_7, cur_block, retdest
+    %rep 8
+        SWAP1
+        DUP2
+        %mstore_kernel_general
+        %increment
+    %endrep
+
+    // stack: addr, cur_block, retdest
+    POP
+    // stack: cur_block, retdest
+    PUSH 0
+    %mload_kernel_general
+    // stack: num_blocks, cur_block, retdest
+    %decrement
+    // stack: num_blocks - 1, cur_block, retdest
+    DUP2
+    // stack: cur_block, num_blocks - 1, cur_block, retdest
+    EQ
+    // stack: is_last_block, cur_block, retdest
+    SWAP1
+    // stack: cur_block, is_last_block, retdest
+    PUSH 1
+    %mload_kernel_general
+    // stack: num_bytes, cur_block, is_last_block, retdest
+
+    // Calculate t counter value.
+    DUP3
+    // stack: is_last_block, num_bytes, cur_block, is_last_block, retdest
+    MUL
+    // stack: is_last_block * num_bytes, cur_block, is_last_block, retdest
+    DUP2
+    // stack: cur_block, is_last_block * num_bytes, cur_block, is_last_block, retdest
+    %increment
+    %mul_const(128)
```

Comment:
this could in theory be its own macro with a semantic name

Reply:
What can? `%mul_const(120)`?